datafusion/expr/src/type_coercion/functions.rs

```diff
@@ -53,20 +54,31 @@ pub fn data_types(
         }
     }
 
-    let valid_types = get_valid_types(&signature.type_signature, current_types)?;
-
+    let mut valid_types = get_valid_types(&signature.type_signature, current_types)?;
     if valid_types
         .iter()
         .any(|data_type| data_type == current_types)
     {
         return Ok(current_types.to_vec());
     }
 
-    // Try and coerce the argument types to match the signature, returning the
-    // coerced types from the first matching signature.
-    for valid_types in valid_types {
-        if let Some(types) = maybe_data_types(&valid_types, current_types) {
-            return Ok(types);
+    // Well-supported signature that returns exact valid types.
+    if !valid_types.is_empty()
+        && matches!(signature.type_signature, TypeSignature::VariadicEqualOrNull)
+    {
+        // exact valid types
+        assert_eq!(valid_types.len(), 1);
+        let valid_types = valid_types.swap_remove(0);
```

Comment:
is it a problem that this may change the type ordering? So that the last type is now the first in the new Vec?

Reply:
I assume there is only one valid types vec, so we always get one. swap_remove it just a fancy way to get the first element in valid_types: Vec<Vec<T>>.

Reply:
equivalent to `valid_types[0].clone` but without clone

Reply:
Ah, got it -- thanks